core/src/banking_stage/decision_maker.rs

```diff
@@ -42,50 +42,40 @@ impl DecisionMaker {
     }
 
     pub(crate) fn make_consume_or_forward_decision(&self) -> BufferedPacketsDecision {
-        let (leader_at_slot_offset, bank_start, would_be_leader, would_be_leader_shortly) = {
-            let poh = self.poh_recorder.read().unwrap();
-            let bank_start = poh
-                .bank_start()
-                .filter(|bank_start| bank_start.should_working_bank_still_be_processing_txs());
-            (
-                poh.leader_after_n_slots(FORWARD_TRANSACTIONS_TO_LEADER_AT_SLOT_OFFSET),
-                bank_start,
-                poh.would_be_leader(HOLD_TRANSACTIONS_SLOT_OFFSET * DEFAULT_TICKS_PER_SLOT),
-                poh.would_be_leader(
-                    (FORWARD_TRANSACTIONS_TO_LEADER_AT_SLOT_OFFSET - 1) * DEFAULT_TICKS_PER_SLOT,
-                ),
-            )
-        };
-
+        let poh_recorder = self.poh_recorder.read().unwrap();
         Self::consume_or_forward_packets(
             &self.my_pubkey,
-            leader_at_slot_offset,
-            bank_start,
-            would_be_leader,
-            would_be_leader_shortly,
+            || {
+                poh_recorder
+                    .bank_start()
+                    .filter(|bank_start| bank_start.should_working_bank_still_be_processing_txs())
+            },
+            || poh_recorder.would_be_leader(HOLD_TRANSACTIONS_SLOT_OFFSET * DEFAULT_TICKS_PER_SLOT),
+            || poh_recorder.would_be_leader(HOLD_TRANSACTIONS_SLOT_OFFSET * DEFAULT_TICKS_PER_SLOT),
+            || poh_recorder.leader_after_n_slots(FORWARD_TRANSACTIONS_TO_LEADER_AT_SLOT_OFFSET),
```

Comment:
each closure is only called if the one above returns a negative (None or false).

So if we have a bank, we immediately return without checking if we would_be_leader

Reply:
Pretty sure original code's intention was to minimize read locking poh by getting all the answers at once, even some of poh calls may not be necessary. Would the refactored code be locking poh too long ?

Reply:
That was actually my intent with this change too, and I believe that it should reduce the time the lock is held.

My bench testing https://github.com/solana-labs/solana/pull/30618#issuecomment-1481908735 showed improvement in the case we have a working bank. I should, however, verify the other cases as well.

Have a meeting, I'll hit you with data-points in a bit.

Reply:
To be clear, in the case where the conditions are all negative there is no expected improvement since we still need to check each.

Reply:
The idea with using these generics fns is that the code (hopefully) becomes equivalent to:

```rust
    pub(crate) fn make_consume_or_forward_decision(&self) -> BufferedPacketsDecision {
        let poh_recorder = self.poh_recorder.read().unwrap();

        if let Some(bank_start) = poh_recorder
            .bank_start()
            .filter(|bank_start| bank_start.should_working_bank_still_be_processing_txs())
        {
            BufferedPacketsDecision::Consume(bank_start)
        } else if poh_recorder
            .would_be_leader(HOLD_TRANSACTIONS_SLOT_OFFSET * DEFAULT_TICKS_PER_SLOT)
        {
            BufferedPacketsDecision::Hold
        } else if poh_recorder
            .would_be_leader(HOLD_TRANSACTIONS_SLOT_OFFSET * DEFAULT_TICKS_PER_SLOT)
        {
            BufferedPacketsDecision::ForwardAndHold
        } else if let Some(x) =
            poh_recorder.leader_after_n_slots(FORWARD_TRANSACTIONS_TO_LEADER_AT_SLOT_OFFSET)
        {
            if x != self.my_pubkey {
                BufferedPacketsDecision::Forward
            } else {
                BufferedPacketsDecision::Hold
            }
        } else {
            BufferedPacketsDecision::Hold
        }
    }
```    


With generics passed, this is what I'd *expect* the compiler to do, and I think its' probably clear why this would result in the lock being held for a short period of time...except potentially in the worst case where we have multiple comparisons.

I should probably modify it so that we drop the lock immediately after calling `leader_pubkey_fn`!

Reply:
> I should probably modify it so that we drop the lock immediately after calling leader_pubkey_fn!

I tried this but there was no noticeable difference